pygfx/objects/_base.py

```diff
@@ -389,153 +406,75 @@ def iter(self, filter_fn=None, skip_invisible=False):
-
-    def get_world_position(self):
-        self.update_matrix_world(update_parents=True, update_children=False)
-        self._v.set_from_matrix_position(self._matrix_world)
-        return self._v.clone()
+    def get_bounding_sphere(self):
+        return la.aabb_to_sphere(self.get_bounding_box())
 
     def get_world_bounding_box(self):
         """Updates all parent and children world matrices, and returns
         a single world-space axis-aligned bounding box for this object's
         geometry and all of its children (recursively)."""
-        self.update_matrix_world(update_parents=True, update_children=True)
-        return self._get_world_bounding_box()
 
-    def _get_world_bounding_box(self):
-        """Returns a world-space axis-aligned bounding box for this object's
-        geometry and all of its children (recursively)."""
-        boxes = []
-        if self._store.geometry:
-            aabb = self._store.geometry.bounding_box()
-            aabb_world = transform_aabb(aabb, self._matrix_world.to_ndarray())
-            boxes.append(aabb_world)
-        if self._children:
-            boxes.extend(
-                [
-                    b
-                    for b in (c.get_world_bounding_box() for c in self._children)
-                    if b is not None
-                ]
-            )
-        if len(boxes) == 1:
-            return boxes[0]
-        if boxes:
-            boxes = np.array(boxes)
-            return np.array([boxes[:, 0].min(axis=0), boxes[:, 1].max(axis=0)])
+        return la.aabb_transform(self.get_bounding_box(), self.world.matrix)
 
     def get_world_bounding_sphere(self):
         """Returns a world-space bounding sphere by converting an
         axis-aligned bounding box to a sphere.
 
         See WorldObject.get_world_bounding_box.
         """
-        aabb = self.get_world_bounding_box()
-        if aabb is not None:
-            return aabb_to_sphere(aabb)
+        return la.aabb_to_sphere(self.get_world_bounding_box())
 
     def _wgpu_get_pick_info(self, pick_value):
         # In most cases the material handles this.
         return self.material._wgpu_get_pick_info(pick_value)
+
+    def look_at(self, target, *, up=None) -> None:
+        if up is None:
+            up = self.up
+
+        position = self.world.position
+        target = np.asarray(target)
+        if np.allclose(target, position):
+            # target and eye are in the same position don't do anything.
+            # Note: the old behavior made it look at (0, 0, 1) in world space
+            # which doesn't seem like a good default
+            return
+
+        new_z = target - position
+        new_z /= np.linalg.norm(new_z)
+        if np.allclose(np.cross(new_z, self.up), 0):
+            # target and up are parallel
+            forward = (0, 0, -1) if self._FORWARD_IS_MINUS_Z else (0, 0, 1)
+            rotation = la.quaternion_make_from_unit_vectors(forward, new_z)
```

Comment:
IIUC, when the view-dir and up-vector or (near) parallel, the view object is rotated towards the new view-dir without considering up, i.e. with an arbitrary up-vector, right? 

I wonder if it makes sense to instead select an alternative up-vector that is at least purely in x, y, or z.

Reply:
To me, there is no axis that intuitively stands out. If the user wants to rotate `forward` onto `up` we don't know which rotation around `new_z` (or rather `up`) they intend to get as the result, so no matter which one we choose it can always be the wrong one.

I was thinking to leave it unconstrained for now and to see if a more intuitive default for our use-cases emerges as we move forward.

Reply:
What makes this feel a bit awkward to me is that this behavior seems not well defined, but we rely on it for the point-light-shadow (for the cameras that point up-down).

Reply:
> What makes this feel a bit awkward to me is that this behavior seems not well defined,

@almarklein It follows the "looking around while maintaining a notion of gravity" philosophy we talked about when you first explained `look_at` to me. If you stood somewhere and "looked up" you would expect to just look up without turning yourself around. If we chose any fixed alternative value for `up` (like (1, 0 ,0)) we would look up, but also turn ourselves around into a fixed position which is typically not the one we started in.

Mathematically, having no default means that we will perform a rotation in the UP-FORWARD-plane of the child frame as long as `up` is a vector in that plane (note that UP in the child frame can be different from the parameter `up`). This is exactly what we wanted (just look up, no rotation around ourselves).

Since, for all WorldObjects (but especially cameras and lights), we start out with local/object-UP, `self.up`, and default `up` being `(0, 1, 0)`, this alignment remains as long as the object's world rotation doesn't have a component rotating around the object's (local) FORWARD axis, i.e., for as long as we can describe its world rotation with an XY-Euler rotation. For lights and cameras this "just works" in 95% of the cases, as we rarely make a camera take a slanted view, and rotating (rotationally symmetric) lights around FORWARD has limited practical use.

If the object's world rotation does include a component around local/object FORWARD then we will still do the right thing so long as `up` remains in the local FORWAD-UP-plane. For example, if we have previously rotated the object using a "normal"/valid call to `look_at(other, up=custom_vector)` then `custom_vector` is now aligned with the local FORWAD-UP-plane. A subsequent call to `look_at(custom_vector, up=custom_vector)` will do the expected thing of "just looking up from where you are now" without adding rotation around local forward.

It breaks when the user expects to have both forward and up rotate at the same time in a well-defined manner. I.e., we can't do "look up while rotating around yourself", because our current signature simply can't express this scenario. Here, bad things can happen, but I think this scenario is rare enough for us to not worry. It is also a bit of a code smell and, now that we have properties for the directions, can be easily resolved by calling `obj.look_at(obj.world.forward, up=new_up)` followed by `obj.look_at(new_up, new_up)`.

Reply:
I think three.js just ever so slightly tilts the up vector just for the duration of the mathematical operation, so that they aren't parallel, and they do so within the math function (analogous to our pylinalg function), not in object3d (analogous to our worldobject class): https://github.com/mrdoob/three.js/blob/4e361e96bc8a1513febb349dde9f1f3be7bf5ec0/src/math/Matrix4.js#L300-L317

This avoids selecting a new "up" vector. Would that strategy help resolve this case?

Reply:
> Mathematically, having no default means that we will perform a rotation in the UP-FORWARD-plane of the child frame as long as up is a vector in that plane

I don't fully follow, but it sounds like the rotation is better defined (i.e. more predictable) then I thought. Since this is what I was mostly worried about (e.g. for the shadow camera) this sounds good enough to me.

Reply:
What is the conclusion from our discussion? Keep the no-default default, or shall I add a small offset to `up_reference` in this case?

Reply:
What would be the consequence of adding a small offset? If I understand correctly it means that the camera will "jump" in terms of rotation about `z_old` as you look at `up_reference`. That seems undesirable.

So I would be OK with keeping the no-default default, I think.